[Merged by Bors] - Handle Geth pre-EIP-155 block sync error condition

jmcph4 opened the pull request:
## Issue Addressed

#2293 

## Proposed Changes

 - Modify the handler for the `eth_chainId` RPC (i.e., `get_chain_id`) to explicitly match against the Geth error string returned for pre-EIP-155 synced Geth nodes
 - ~~Add a new helper function, `rpc_error_msg`, to aid in the above point~~
 - Refactor `response_result` into `response_result_or_error` and patch reliant RPC handlers accordingly (thanks to @pawanjay176)

## Additional Info

Geth, as of Pangaea Expanse (v1.10.0), returns an explicit error when it is not synced past the EIP-155 block (2675000). Previously, Geth simply returned a chain ID of 0 (which was obviously much easier to handle on Lighthouse's part).

jmcph4 committed:
Handle Geth pre-EIP-155 block sync error condition

Geth, as of Pangaea Expanse (v1.10.0), returns an explicit error when
it is not synced past the EIP-155 block (2675000). Previously, Geth
simply returned a chain ID of 0 (which was obviously much easier to
handle on Lighthouse's part).

This commit explicitly handles this error condition by modifying the
`get_chain_id` function in `beacon_node/eth1/src/http.rs` to `match`
against the error string returned by Geth.

CLAassistant commented:
[![CLA assistant check](https://cla-assistant.io/pull/badge/signed)](https://cla-assistant.io/sigp/lighthouse?pullRequest=2304) <br/>All committers have signed the CLA.

jmcph4 commented:
Marking this as draft currently for a few reasons:

 - Confirming CI is happy
 - Likely prettier Rust that could be used
 - First time contributing to Lighthouse

jmcph4 commented:
I've been considering whether or not to tidy this up a bit but I feel it's overengineering at this point. Likely out of scope for this PR, but should we add types for handling [JSON-RPC 2.0 errors](https://www.jsonrpc.org/specification#error_object)?

jmcph4 committed:
Add `rpc_error_msg` helper function

This commit adds a helper function, `rpc_error_msg`, which extracts the
human-readable error message from a JSON-RPC error message. This is
done to aid the handling of EIP-155-related errors from Geth inside the
`get_chain_id` RPC handler.

jmcph4 committed:
Use `nth(1)` rather than `skip(1).next()`

This commit fixes Clippy complaints due to usage of `skip(1).next()`
instead of the much simpler `nth(1)`.

pawanjay176 commented on beacon_node/eth1/src/http.rs:
```diff
@@ -423,6 +439,29 @@ fn response_result(response: &str) -> Result<Option<Value>, String> {
+///
+/// Note: the input string, `s`, will be of the form:
+///     "Eth1 node returned error: {\"code\":[code],\"message\":\"[message]\"}"
+fn rpc_error_msg(s: &str) -> Result<String, String> {
```
Working on the error string here by manually splitting the json string seems a little convoluted. Instead you can directly work on the json object by modifying the `response_result` function to return the error string itself. Something like
```rust
/// Accepts an entire HTTP body (as a string) and returns either the `result` field or the `error['message']` field, as a serde `Value` .
fn response_result_or_error(response: &str) -> Result<Result<Value, Value>, String> {
    let json = serde_json::from_str::<Value>(&response)
        .map_err(|e| format!("Failed to parse response: {:?}", e))?;

    if let Some(error) = json.get("error").map(|e| e.get("message")).flatten() {
        Ok(Err(error.clone()))
    } else {
        let res = json
            .get("result")
            .cloned()
            .ok_or_else(|| "Failed to get result".to_string())?;
        Ok(Ok(res))
    }
}
```

Then in the `get_chain_id` function, you can check the error string from the above function.

jmcph4 replied:
This is much more elegant, provided we're happy with modifying the other handlers to accommodate the changed function signature?

jmcph4 replied:
Actually I think this is a pretty clean change, consult e77a41a.

pawanjay176 reviewed (changes requested):
This is great work! I had a suggestion to reduce manual string operations on the json string. Let me know what you think :)

jmcph4 committed:
Refactor RPC error handling

This commit refactors how RPC error responses are handled by making the
following changes:

 - The helper `response_result` is now `response_result_or_error` and
   handles RPC error responses directly (thanks to @pawanjay176)
 - Each RPC handler relying on `response_result` has now been patched
   to support the new API (mostly by inserting an `ok()` call
   immediately after each `response_result_or_error` call in order to
   preserve the `Option`-based expectations)
 - The previously-added helper, `rpc_error_msg` has been removed as its
   now redundant

pawanjay176 commented on beacon_node/eth1/src/http.rs:
```diff
@@ -91,14 +92,29 @@ pub async fn get_network_id(endpoint: &str, timeout: Duration) -> Result<Eth1Id,
+
+    /* extract response text here */
+    let response_result: Result<Result<Value, Value>, String> =
+        response_result_or_error(&response_body);
```
Can directly return the result here so you don't have to handle the outer error case separately
```suggestion
        response_result_or_error(&response_body)?;
```

jmcph4 replied:
Thanks again for this! I should've checked the return types more thoroughly. Consult 867afdb.

pawanjay176 reviewed (approved):
Looks great. Just a minor nitpick :)

jmcph4 committed:
Simplify error handling logic in `get_chain_id`

This commit simplifies how errors are handled in the `get_chain_id`
handler by using `?` on the initial call to `response_result_or_error`.

Once again, thanks to @pawanjay176!

jmcph4 committed:
Fix formatting

This commit fixes a missed formatting issue.

paulhauner commented on beacon_node/eth1/src/http.rs:
```diff
@@ -91,14 +92,22 @@ pub async fn get_network_id(endpoint: &str, timeout: Duration) -> Result<Eth1Id,
+    let response_body: String =
+        send_rpc_request(endpoint, "eth_chainId", json!([]), timeout).await?;
+
+    /* extract response text here */
```
It's nitpick, but comments in the LH codebase generally look like this:

```suggestion
    // Extract response text here.
```

jmcph4 replied:
Consult 31be1ab.

paulhauner commented on beacon_node/eth1/src/http.rs:
```diff
@@ -91,14 +92,22 @@ pub async fn get_network_id(endpoint: &str, timeout: Duration) -> Result<Eth1Id,
+    /* extract response text here */
+    let response_result: Result<Value, Value> = response_result_or_error(&response_body)?;
+
+    /* specifically handle Geth's pre-EIP-155 sync error message */
```
As per previous comment nitpick.

jmcph4 replied:
Also consult 31be1ab.

paulhauner commented on beacon_node/eth1/src/http.rs:
```diff
@@ -407,19 +421,19 @@ pub async fn send_rpc_request(
 
-/// Accepts an entire HTTP body (as a string) and returns the `result` field, as a serde `Value`.
-fn response_result(response: &str) -> Result<Option<Value>, String> {
+/// Accepts an entire HTTP body (as a string) and returns either the `result` field or the `error['message']` field, as a serde `Value` .
```
```suggestion
/// Accepts an entire HTTP body (as a string) and returns either the `result` field or the `error['message']` field, as a serde `Value`.
```

jmcph4 replied:
@paulhauner As always, nitpicks are welcome! These are all valid changes to want for the codebase. 

Currently waiting on CI, although I don't see what would regress it.

paulhauner reviewed (changes requested):
Awesome, thanks for getting involved @jmcph4. Glad to have your contributions :rocket: 

I've added a couple of nitpicks and a suggestion about the `response_result_or_error`. I know it conflicts with @pawanjay176, so both of you please let me know if I'm missing the mark :)

jmcph4 committed:
Fix comment styling

This commit fixes the styling on comments left in `http.rs` in order to
better comply with the style guide. Now, C-style comments have been
converted to their more C++-like equivalents and some minor punctuation
fixes have been applied.

Thanks to @paulhauner for pointing these issues out - I certainly
missed them!

paulhauner commented:
> I've added a couple of nitpicks and a suggestion about the response_result_or_error

It seems like my comment regarding this was lost :( I'll rewrite it in the coming day(s).

paulhauner commented on beacon_node/eth1/src/http.rs:
```diff
@@ -407,19 +421,19 @@ pub async fn send_rpc_request(
-/// Accepts an entire HTTP body (as a string) and returns the `result` field, as a serde `Value`.
-fn response_result(response: &str) -> Result<Option<Value>, String> {
+/// Accepts an entire HTTP body (as a string) and returns either the `result` field or the `error['message']` field, as a serde `Value`.
+fn response_result_or_error(response: &str) -> Result<Result<Value, Value>, String> {
```
I'm not sure about this `Result<Result<Value, Value>, String>`, it feels like an anti-pattern and I think the present implementation will result in some less-than-ideal errors.

For example, if we get an `"error"` field on `fn get_deposit_logs_in_range` we'll log an error saying `"No result field was returned for deposit logs"`, when it would be ideal to indicate that we got an error and the value of it.

I'd suggest something like this:

```rust
#[derive(Debug)]
enum RpcError {
    NoResultField,
    Error(String)
}

fn response_result_or_error(response: &str) -> Result<Value, RpcError> {
    ...
}
```
And then in `fn get_deposit_logs_in_range` we can do:

```rust
    response_result_or_error(&response_body)
        .map_err(|e| format!("Unable to obtain deposit logs: {:?}", e)?
```

jmcph4 replied:
Making an explicit error type would be better, agreed.

While it might be out of scope for this PR specifically, I feel like this gets back to [my comment](https://github.com/sigp/lighthouse/pull/2304#issuecomment-815539934) about supporting the entirety of the set of JSON-RPC 2.0 error conditions? After all, it seems like that's what Geth is sending to us anyway.

paulhauner replied:
> I feel like this gets back to my comment about supporting the entirety of the set of JSON-RPC 2.0 error conditions? After all, it seems like that's what Geth is sending to us anyway.

I don't feel so strongly either way about parsing the `error` struct instead of flattening it to a `String`. AFAIK, ultimately we just flatten it and present it to the user as a `String` anyway, so I'm not sure there's a whole lot to gain.

jmcph4 replied:
> AFAIK, ultimately we just flatten it and present it to the user as a String anyway, so I'm not sure there's a whole lot to gain.

I think we err on the side of simplicity and just pass the string on, so I'm inclined to agree 
here.

I'll work on your proposed changes now.

paulhauner reviewed (changes requested):
Thanks for the speedy response to my other comments. Here's that comment that got lost along the way, apologies for the delay.

jmcph4 committed:
Simplify handling of RPC messages

This commit simplifies how RPC messages are handled by defining a new
type, `RpcError`, and making `response_result_or_result` return a
single `Result` with `RpcError` as its error type.

jmcph4 committed:
Modify `get_deposit_logs_in_range` to pass on errors

This commit makes the `get_deposit_logs_in_range` function pass on
error strings derive from `RpcError` types it receives from
`response_result_or_error`.

jmcph4 commented:
@paulhauner Sorry this latest iteration took so long, hopefully the current changes are what you had in mind.

It appears that CI jobs aren't running for this anymore as I'm a first-time contributor? Strange given they've run fine every other time, but it might be me misunderstanding whatever the GitHub UI is trying to tell me.

paulhauner commented:
Hi @jmcph4, sorry for the delay here. We've got a big PR backlog at the moment with Altair and Rayonism running in parallel. We're making good progress though and are coming back to PR reviews.

Would you mind fixing the CI conflict here please? I approved your CI run :)

jmcph4 committed:
Bump deps to mitigate upstream vulnerabilities

This commit applies the changes from `cargo update` to the lockfile
in order to mitigate two vulnerabilities in upstream dependencies,
namely:

 - [RUSTSEC-2021-0056](https://rustsec.org/advisories/RUSTSEC-2021-0056)
 - [RUSTSEC-2021-0055](https://rustsec.org/advisories/RUSTSEC-2021-0055)

jmcph4 commented:
@paulhauner dbe15afe fixes my `make audit` locally, but it appears that CI is only running the CLA check again.

As for the linting failure, `make lint` works for me locally, but I'll continue investigating.

paulhauner commented:
> As for the linting failure, `make lint` works for me locally, but I'll continue investigating.

This is usually fixed by a `rustup update`, might be worth a shot :)

jmcph4 committed:
Merge branch 'unstable' into jmcph4-2293-unsynced-eth1-eip155-errors

jmcph4 commented:
@paulhauner CI should be clear now and provided you're happy with the actual changeset I think this might be good to go? Let me know if I've missed anything (sorry this slid out for so long).

paulhauner commented on beacon_node/eth1/src/http.rs:
```diff
@@ -48,6 +48,30 @@ pub enum BlockQuery {
+#[derive(Debug, Serialize, Deserialize)]
+pub enum RpcError {
+    NoResultField,
+    InvalidJson,
```
We've lost the ability to see the exact JSON error here. Sometimes `serde` gives useful information that can narrow down the error. Making this `InvalidJson(String)` would ensure we retain the error.

paulhauner commented on beacon_node/eth1/src/http.rs:
```diff
@@ -282,7 +324,7 @@ async fn call(
 
     let response_body = send_rpc_request(endpoint, "eth_call", params, timeout).await?;
-    match response_result(&response_body)? {
+    match response_result_or_error(&response_body).ok() {
```
We're suppressing errors here.

Previously, if the server returned an `error` in the response we'd exit early with `Err`. Now, we return `Ok(None)`, indicating the server successfully returned with no data.

paulhauner commented on beacon_node/eth1/src/http.rs:
```diff
@@ -48,6 +48,30 @@ pub enum BlockQuery {
+impl From<RpcError> for String {
+    fn from(value: RpcError) -> Self {
+        match value {
+            RpcError::NoResultField => "No JSON error returned".into(),
```
This would probably be more helpful as somthing like: "No result field in JSON response"

paulhauner commented on beacon_node/eth1/src/http.rs:
```diff
@@ -141,22 +180,24 @@ pub async fn get_block(
-        response_result(&response_body)?
+    let hash: Vec<u8> = hex_to_bytes(
+        response_result_or_error(&response_body)
+            .ok()
```
When we use `ok()` here we're still flattening all the different error types into "No result field was returned for block".

So, if there's an `RpcError::Error(s)` we're still presenting it to the user as if there was no result field.

paulhauner commented on beacon_node/eth1/src/http.rs:
```diff
@@ -141,22 +180,24 @@ pub async fn get_block(
     let timestamp = hex_to_u64_be(
-        response_result(&response_body)?
+        response_result_or_error(&response_body)
+            .ok()
```
As above regarding flattening errors.

paulhauner commented on beacon_node/eth1/src/http.rs:
```diff
@@ -165,7 +206,8 @@ pub async fn get_block(
     let number = hex_to_u64_be(
-        response_result(&response_body)?
+        response_result_or_error(&response_body)
+            .ok()
```
As above regarding flattening errors.

paulhauner commented on beacon_node/eth1/src/http.rs:
```diff
@@ -322,7 +364,8 @@ pub async fn get_deposit_logs_in_range(
     let response_body = send_rpc_request(endpoint, "eth_getLogs", params, timeout).await?;
-    response_result(&response_body)?
+    Ok(response_result_or_error(&response_body)
+        .ok()
```
As above regarding flattening errors.

paulhauner committed:
Apply review comments

paulhauner reviewed (changes requested):
Thanks for persisting with this and apologies for the slow turn-around times.

There's a few changes I'd like to request. In the interests of giving feedback, I left a comment for each change request. In the interests of saving your time (and allowing me to understand it better) I created a commit which contains my recommendations:

https://github.com/paulhauner/lighthouse/commit/384b35413826d4fdaf2ddab28b73223bd0f8b04a

Feel free to just cherry-pick this commit onto your branch, if you agree with them.

jmcph4 commented:
> Thanks for persisting with this and apologies for the slow turn-around times.
> 
> There's a few changes I'd like to request. In the interests of giving feedback, I left a comment for each change request. In the interests of saving your time (and allowing me to understand it better) I created a commit which contains my recommendations:
> 
> [paulhauner@384b354](https://github.com/paulhauner/lighthouse/commit/384b35413826d4fdaf2ddab28b73223bd0f8b04a)
> 
> Feel free to just cherry-pick this commit onto your branch, if you agree with them.

@paulhauner Thanks for the feedback, it's much appreciated. I've reviewed your commit and cherry-picked it into this branch as you suggested. Apologies that this wasn't as smooth as it could've been.

paulhauner commented:
> Thanks for the feedback, it's much appreciated. I've reviewed your commit and cherry-picked it into this branch as you suggested. Apologies that this wasn't as smooth as it could've been.

Thanks for your contribution, it's much appreciated. It's great to see people digging in and fixing annoyances :heart: 

We're waiting to clear the v1.4.0 release. Once that's done, we can merge this and include it in v1.5.0 :)

paulhauner reviewed (approved):
Let's get this merged! Thanks again @jmcph4.

bors r+

paulhauner commented:
bors r-

paulhauner commented:
bors r+
